src/alpaka/plugin-SiPixelClusterizer/gpuClusterChargeCut.h

```diff
@@ -0,0 +1,147 @@
+      const uint32_t blockIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u]);
+      if (blockIdx >= moduleStart[0])
+        return;
+
+      auto firstPixel = moduleStart[1 + blockIdx];
+      auto thisModuleId = id[firstPixel];
+      assert(thisModuleId < MaxNumModules);
+      assert(thisModuleId == moduleId[blockIdx]);
+
+      auto nclus = nClustersInModule[thisModuleId];
+      if (nclus == 0)
+        return;
+
+      const uint32_t threadIdxLocal(alpaka::idx::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
+      if (threadIdxLocal == 0 && nclus > MaxNumClustersPerModules)
+        printf("Warning too many clusters in module %d in block %d: %d > %d\n",
+               thisModuleId,
+               blockIdx,
+               nclus,
+               MaxNumClustersPerModules);
+
+      // Stride = block size.
+      const uint32_t blockDimension(alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u]);
+
+      // Get thread / CPU element indices in block.
+      const auto& [firstElementIdxNoStride, endElementIdxNoStride] =
+          cms::alpakatools::element_index_range_in_block(acc, Vec1::all(firstPixel));
+
+      if (nclus > MaxNumClustersPerModules) {
+        uint32_t firstElementIdx = firstElementIdxNoStride[0u];
+        uint32_t endElementIdx = endElementIdxNoStride[0u];
+        // remove excess  FIXME find a way to cut charge first....
+        for (uint32_t i = firstElementIdx; i < numElements; ++i) {
+          if (!cms::alpakatools::get_next_element_1D_index_stride(
+                  i, firstElementIdx, endElementIdx, blockDimension, numElements))
+            break;
+          if (id[i] == InvId)
+            continue;  // not valid
+          if (id[i] != thisModuleId)
+            break;  // end of module
+          if (clusterId[i] >= MaxNumClustersPerModules) {
+            id[i] = InvId;
+            clusterId[i] = InvId;
+          }
+        }
+        nclus = MaxNumClustersPerModules;
+      }
+
+#ifdef GPU_DEBUG
+      if (thisModuleId % 100 == 1)
+        if (threadIdxLocal == 0)
+          printf("start clusterizer for module %d in block %d\n", thisModuleId, blockIdx);
+#endif
+
+      auto&& charge = alpaka::block::shared::st::allocVar<int32_t[MaxNumClustersPerModules], __COUNTER__>(acc);
+      auto&& ok = alpaka::block::shared::st::allocVar<uint8_t[MaxNumClustersPerModules], __COUNTER__>(acc);
+      auto&& newclusId = alpaka::block::shared::st::allocVar<uint16_t[MaxNumClustersPerModules], __COUNTER__>(acc);
+
+      assert(nclus <= MaxNumClustersPerModules);
+      cms::alpakatools::for_each_element_1D_block_stride(acc, nclus, [&](uint32_t i) { charge[i] = 0; });
+      alpaka::block::sync::syncBlockThreads(acc);
+
+      uint32_t firstElementIdx = firstElementIdxNoStride[0u];
+      uint32_t endElementIdx = endElementIdxNoStride[0u];
+      for (uint32_t i = firstElementIdx; i < numElements; ++i) {
+        if (!cms::alpakatools::get_next_element_1D_index_stride(
```

Comment:
Yes the issue here was that 2 loops are introduced with Alpaka: one for the strided access, and an inner one for the CPU elements. Whether using a helper function and a lambda, or directly doing the looping on call site, one faces this 2-loops situation.
This was an issue when the legacy code had break statements, since obviously only the innermost loop was exited. A trivial way around I had done first, was of course to use a boolean, to exit the innermost loop when needed.
But in a sense, the gymnastics with the boolean had to be repeated each time there was suck a block on call site, hence code duplication.

This helper function allows to have the same logic as 2 loops but with only 1 loop and no extra storage, which makes the portability of the legacy code easier, as nothing has to be changed regarding break statements.